File: Greedy&Dynamic/CoinChange.c
#include<stdio.h>

int coinChange(int coins[], int n, int amount) {
    int dp[amount + 1];
    dp[0] = 0;

    for (int i = 1; i <= amount; i++) {
        dp[i] = amount + 1; // Initialize with a value greater than the maximum amount
        for (int j = 0; j < n; j++) {
            if (coins[j] <= i) {
                dp[i] = (dp[i] < 1 + dp[i - coins[j]]) ? dp[i] : 1 + dp[i - coins[j]];
            }
        }
    }

    return (dp[amount] > amount) ? -1 : dp[amount];
}

int main() {
    int coins[] = {1, 2, 5};
    int n = sizeof(coins) / sizeof(coins[0]);
    int amount = 11;
    int minCoins = coinChange(coins, n, amount);
    if (minCoins == -1) {
        printf("It is not possible to make the amount using given coins.\n");
    } else {
        printf("Minimum number of coins required to make the amount is: %d\n", minCoins);
    }
    return 0;
}

File: Strivers A2Z DSA/Greedy/Coin_Change.java
public class Coin_Change 
{
    public int coinChange(int[] coins, int amount) 
    {
        int n = coins.length;

        for(int i=0;i<n-1;i++)
        {
            for(int j=0;j<n-1;j++)
            {
                if(coins[j]<coins[j+1])
                {
                    int temp = coins[j];
                    coins[j] = coins[j+1];
                    coins[j+1] = temp;
                }
            }
        }

        int num = 0;
        int remaining = amount;
        for(int i=0;i<n;i++)
        {
            if(remaining/coins[i] > 0)
            {
                num += remaining/coins[i];
                remaining -= coins[i]*(remaining/coins[i]);
            }
        }

        if (remaining != 0) 
            return -1;
        return num;
    }    
}
